### DISCUSSION TITLE
Compile Issues

### DISCUSSION THREAD

**Discussion by ian-hargreaves** [astlouys/Pico-Green-Clock#33] [General]
Hi there
I'm new, but learning fast to get Green Clock running
I have the basics sorted and can compile the original source code for the clock and deploy my own UF2

I've downloaded this repository and get errors trying to compile.
I have a basic Pico so copied CMakeLists.txt.Pico to CMakeLists.txt

CMAKE worrks, but MAKE -j4 fails. 
It looks like its trying to reference the Pico W version or I have some files missing from my setup? 

Thank you to all who have helped this project. 
Much appreciated.
Ian

[ 13%] Built target picotoolBuild
[ 16%] Building C object CMakeFiles/Pico-Green-Clock.dir/Pico-Green-Clock.c.o
[ 16%] Building C object CMakeFiles/Pico-Green-Clock.dir/home/ian-h/pico/pico-sdk/src/rp2_common/hardware_gpio/gpio.c.o
[ 16%] Building C object CMakeFiles/Pico-Green-Clock.dir/Ds3231.c.o
[ 16%] Building C object CMakeFiles/Pico-Green-Clock.dir/home/ian-h/pico/pico-sdk/src/rp2_common/pico_stdlib/stdlib.c.o
/home/ian-h/pico/Pico-Green-Clock-main/Pico-Green-Clock.c:313:2: warning: #warning Built as RELEASE_VERSION [-Wcpp]
  313 | #warning Built as RELEASE_VERSION
      |  ^~~~~~~
[ 17%] Building C object CMakeFiles/Pico-Green-Clock.dir/home/ian-h/pico/pico-sdk/src/rp2040/pico_platform/platform.c.o
In file included from /home/ian-h/pico/Pico-Green-Clock-main/Pico-Green-Clock.c:811:
/home/ian-h/pico/Pico-Green-Clock-main/picow_ntp_client.h:11:10: fatal error: pico/cyw43_arch.h: No such file or directory
   11 | #include "pico/cyw43_arch.h"

**Comment by ian-hargreaves** [comment]
OK - I think I fixed this ... 
Pico-Green-Clock.c 
comment out the line : 
#define PICO_W  ///
becomes
// #define PICO_W  ///

**Comment by ian-hargreaves** [comment]
And if I had read the manual properly ... I would have seen this: 
Page 100: 
_The same source code supports both the Pico and the Pico W for NTP support. All you
have to do is comment (or uncomment) the #define PICO_W in the source code to enable
Pico W (and NTP) support. (Take note that the #define NTP_ENABLE has not been
implemented yet and it is the #define PICO_W that controls the NTP support code for
now._

Anyway - I have a Pico W arriving so will be swapping to that to get NTP support shortly.

**Reply by astlouys** [reply]
Hello Ian,

It looks like you found the problem. Congratulations !

I could have tuned the CMakeLists.txt to make it the only thing to modify to differentiate the Pico VS PicoW, but I thought it could happen, for some reason, that someone using a PicoW may want to turn Off NTP (and by the way, the DS3231 proved to keep a very good precision on the long term). You will appreciate the PicoW. If you want to go a step further, take a look at the Pico-RGB-Matrix (and the corresponding Firmware in my repository). It makes a good start for a Smart Home system.

May I ask you in which country you are ?

André

**Reply by astlouys** [reply]
Hello Ian,

It looks like you found the problem. Congratulations !

I could have tuned the CMakeLists.txt to make it the only thing to modify to differentiate the Pico VS PicoW, but I thought it could happen, for some reason, that someone using a PicoW may want to turn Off NTP (and by the way, the DS3231 proved to keep a very good precision on the long term). You will appreciate the PicoW. If you want to go a step further, take a look at the Pico-RGB-Matrix (and the corresponding Firmware in my repository). It makes a good start for a Smart Home system.

May I ask you in which country you are ?

André

**Comment by ian-hargreaves** [comment]
Hi André

Many thanks for the quick reply.

I'm in the UK, England.

I saw the Pico-RBG-Matrix, might take a look at that ... For now, I like the simplicity of the Green Clock.
I used to be a decent C programmer 30 years ago and can remember most of the basics ... So might have a go at disabling the blinking dots as a test of my C coding skills.
Or even add a new option for the user to control the blinking dots.

Are you still updating this project yourself ?

If I do make changes - is there a process to get them back to you to consider for including in your project ?

Thanks
Ian

________________________________

**Reply by astlouys** [reply]
Hi Ian,

You're right, the Green Clock is relatively simple to program and use... and don't worry, programming is like riding a bicycle: you never really forget :-)  You may only need a few days / weeks to refresh.

It's been a while since the last update I made to the Green Clock. In fact, I had a system crash after putting version 10.00 online. Also, I spent the last months on the RGB Matrix. The larger display (64 X 32), colors and remote control allow it to go on step further than the Green Clock for the potential features. I didn't update GitHub recently for the RGB Matrix since I haven't seen much interest, but will eventually do it in the future.

As for updating the Green Clock, I could do it. Since I'm not familiar with all the details of GitHub, if you can provide me with the "diff" between the original and your update, I could add them, as long as they could be considered "public interest". I know of one bug: if your WiFi gets down for some reason, the Green Clock will not reconnect after it is back (for NTP). This is something I should eventually take care of in a next version.

André

**Comment by ian-hargreaves** [comment]
I'm trying to compile the source for my Pico_W ... 
I switched the cmake files, put the PICO_W line back in the #define in the source. 
cmake .. runs OK. 
make -j fails with ... 

home/ian-h/pico/Pico-Green-Clock-main/picow_ntp_client.c:21:10: fatal error: lwip/dns.h: No such file or directory
   21 | #include "lwip/dns.h"

I've tried for a couple of hours to google answers and not found something that fixes it for me ... 
Any thoughts? 

Guessing some stnadard libraries are not being included or some environment variable is missing. 
I have set env for PICO_SDK_PATH
and PICO_BOARD

**Reply by ian-hargreaves** [reply]
OK - I fixed this ... in the CMakeLists.txt move the line set(PICO_BOARD pico_w) to ABOVE the include for the SDK
Like this: 

set(PICO_BOARD pico_w)
include(pico_sdk_import.cmake)

**Reply by astlouys** [reply]
Which SDK version are you using ? You may want to temporarily change the SDK version to see if it helps.

I did face a similar situation while updating the SDK after a system crash and spent days before getting  back on track. I had a few problems at the same time, so it's been difficult for me to pinpoint what was the source of which problem, but switching SDK versions eventually helps.

Also, what is your development environment ? Raspberry Pi 5 (or Windows) ?  Visual Studio Code ?

**Reply by astlouys** [reply]
Which SDK version are you using ? You may want to temporarily change the SDK version to see if it helps.

I did face a similar situation while updating the SDK after a system crash and spent days before getting  back on track. I had a few problems at the same time, so it's been difficult for me to pinpoint what was the source of which problem, but switching SDK versions eventually helps.

Also, what is your development environment ? Raspberry Pi 5 (or Windows) ?  Visual Studio Code ?

**Comment by astlouys** [comment]
Our messages were sent  at the same time... Great that you find the solution !

**Comment by ian-hargreaves** [comment]
And in case someone else thinks there's a problem - switching to the Pico W I notice a much longer start up time for the Pico. The green LED flashes for a long time. 
I thought something wasn't working - it was .. it just takes a while to startup.

**Reply by astlouys** [reply]
Good point, Ian,

In fact, the green LED (on the PicoW board itself) indicates the current retry number trying to connect to the local WiFi network (and a train of fast blink at the end  of the retries in case of a failure). Of course, the network credentials are not properly configured the first time you start the Firmware (network credentials are simply defaults which, of course do not correspond to your WiFi setup). So, it takes some time for the PicoW to give up, as indicated in the User Guide (in my version of the Guide, it is in the section "About Version xxx", at the very beginning, just before the "*** IMPORTANT ***" notice).

Whenever the WiFi credentials are properly setup, the startup sequence should be much quicker (sometimes, a wrong NTP server address is returned and in such cases, the delay may be a little longer).

**Reply by ian-hargreaves** [reply]
This make sense ... I set Wifi Credentials in the source code ... but I didn't change the NTP settings. I have NTP running on my NAS so I need to check that.

**Reply by ian-hargreaves** [reply]
BTW - What's the easiest way to get the web interface working that someone else added in a fork of this code? 
Apologies if this is obvious - I've never done this before.

**Reply by astlouys** [reply]
I'm not too familiar with GitHub either. I guess there must be a way to directly download and use the forked code version...

You may want to contact cmcornish (on GitHub) to check with him how this can be done. At some point, I thought about updating my source code, but as I mentioned before, I'm now developing the Pico-RGB-Matrix which offers much more features on which to build a smart home system and I still have plans for so many things on it that a Green Clock update definitely won't be in the short term :-)
